File: dongdaang/dfs_bfs/Shortest_Bridge.py
#https://leetcode.com/problems/shortest-bridge/


#1. dfs로 2개의 대륙 중 하나만 찾음
#2. 찾은 대륙에서 bfs를 통해 또 다른 대륙까지의 거리 측정

from collections import deque

class Solution:
    def shortestBridge(self, grid):
        N = len(grid)
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]
        queue = deque()
        
        def dfs(x, y):
            grid[x][y] = -1         #dfs를 수행할 때, 이미 방문한 위치는 -1로 변경
            queue.append((x, y))
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < N and 0 <= ny < N and grid[nx][ny] == 1:
                    dfs(nx, ny)
                    
        def bfs():
            while queue:
                x, y = queue.popleft()
                for i in range(4):
                    nx = x + dx[i]
                    ny = y + dy[i]
                    if nx < 0 or nx >= N or ny < 0 or ny >= N or grid[nx][ny] == -1:        #이동할 위치가 범위 밖이거나 해당 값이 -1인 경우 continue
                        continue   
                    if grid[nx][ny] != 1 and grid[nx][ny] != 0: #도착지이가 아니거나 0이 아니면 방문할 필요 없으므로 continue
                        continue
                    if grid[nx][ny] == 1:                       #값이 1이면 다른 대륙에 도착한 것을 의미하므로 값을 리턴
                        return abs(grid[x][y]) - 1              #거리 값을 빼주면서 전진했기 때문에 반환 값은 절댓값(해당 위치 값) - 1
                    elif grid[nx][ny] == 0:                     #값이 0이면 계속 진행
                        grid[nx][ny] = grid[x][y] - 1           #거리 값을 빼주면서 전진
                        queue.append((nx, ny))

        for i in range(N):
            for j in range(N):
                if grid[i][j] == 1:
                    dfs(i, j)           #dfs를 한번만 수행하고 바로 bfs 수행
                    return bfs()

a = Solution()
print(a.shortestBridge([[0,1,0],[0,0,0],[0,0,1]]))

File: CokeLee777/src/com/leetcode/dfsandbfs/ShortestBridge.java
package com.leetcode.dfsandbfs;

import java.util.LinkedList;
import java.util.Queue;

/**
 * https://leetcode.com/problems/shortest-bridge/
 */
public class ShortestBridge {

    public static class MySourceCode{

        private final int[] dx = {0,0,1,-1};
        private final int[] dy = {1,-1,0,0};
        private int islandNum = 1;

        public int shortestBridge(int[][] grid) {
            Queue<int[]> queue = new LinkedList<>();
            int[][] bridge = new int[grid.length][grid[0].length];          //최단거리 테이블
            boolean[][] visited = new boolean[grid.length][grid[0].length]; //방문처리 배열

            //dfs로 섬의 번호 매기기
            for(int i = 0; i < grid.length; i++){
                for(int j = 0; j < grid[i].length; j++){
                    if(grid[i][j] == 1 && !visited[i][j]){
                        dfs(grid, visited, i, j, queue);
                        islandNum++;
                    }
                }
            }

            //bfs로 1번섬과 2번섬과의 최단거리 측정
            int min = Integer.MAX_VALUE;
            while(!queue.isEmpty()){
                int[] now = queue.poll();

                for(int i = 0; i < 4; i++){
                    int nx = now[0] + dx[i];
                    int ny = now[1] + dy[i];

                    if(nx < 0 || nx >= grid.length || ny < 0 || ny >= grid[0].length)
                        continue;
                    //섬이 아니라면 다리를 세운다
                    if(grid[nx][ny] != 1 && grid[nx][ny] != 2){
                        //이미 세워진 다리라면 지금 경로의 최단거리와 비교
                        if(bridge[nx][ny] != 0){
                            bridge[nx][ny] = Math.min(bridge[nx][ny], bridge[now[0]][now[1]] + 1);
                        }//처음 세운 다리라면 지금 거리 + 1을 넣고 큐에 삽입
                        else {
                            queue.offer(new int[]{nx, ny});
                            bridge[nx][ny] = bridge[now[0]][now[1]] + 1;
                        }
                    }//또다른 섬이라면 모든 경로로부터 최단거리 비교
                    else if(grid[nx][ny] == 2){
                        min = Math.min(min, bridge[now[0]][now[1]]);
                    }
                }
            }

            return min;
        }

        private void dfs(int[][] grid, boolean[][] visited, int x, int y, Queue<int[]> queue){
            //방문처리와 섬번호 처리
            visited[x][y] = true;
            grid[x][y] = islandNum;
            //섬번호가 1인 것만 큐에 삽입
            if(islandNum == 1) queue.offer(new int[]{x, y});

            for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(nx < 0 || nx >= grid.length || ny < 0 || ny >= grid[0].length)
                    continue;
                if(grid[nx][ny] == 1 && !visited[nx][ny]){
                    dfs(grid, visited, nx, ny, queue);
                }
            }
        }
    }

    public static class Solution{

        public int shortestBridge(int[][] A) {
            int m = A.length, n = A[0].length;
            boolean[][] visited = new boolean[m][n];
            int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
            Queue<int[]> q = new LinkedList<>();
            boolean found = false;
            // 1. dfs to find an island, mark it in `visited`
            for (int i = 0; i < m; i++) {
                if (found) {
                    break;
                }
                for (int j = 0; j < n; j++) {
                    if (A[i][j] == 1) {
                        dfs(A, visited, q, i, j, dirs);
                        found = true;
                        break;
                    }
                }
            }
            // 2. bfs to expand this island
            int step = 0;
            while (!q.isEmpty()) {
                int size = q.size();
                while (size-- > 0) {
                    int[] cur = q.poll();
                    for (int[] dir : dirs) {
                        int i = cur[0] + dir[0];
                        int j = cur[1] + dir[1];
                        if (i >= 0 && j >= 0 && i < m && j < n && !visited[i][j]) {
                            if (A[i][j] == 1) {
                                return step;
                            }
                            q.offer(new int[]{i, j});
                            visited[i][j] = true;
                        }
                    }
                }
                step++;
            }
            return -1;
        }
        private void dfs(int[][] A, boolean[][] visited, Queue<int[]> q, int i, int j, int[][] dirs) {
            if (i < 0 || j < 0 || i >= A.length || j >= A[0].length || visited[i][j] || A[i][j] == 0) {
                return;
            }
            visited[i][j] = true;
            q.offer(new int[]{i, j});
            for (int[] dir : dirs) {
                dfs(A, visited, q, i + dir[0], j + dir[1], dirs);
            }
        }
    }
}
